docs/proposals/kubelet-systemd.md

```diff
@@ -0,0 +1,277 @@
+
+Classical `cgroup` hierarchies were orthogonal per resource group controller, and a process could
+exist in different parts of the hierarchy.
+
+For example, a process `p1` could exist in each of the following at the same time:
+
+* `/sys/fs/cgroup/cpu/important/`
+* `/sys/fs/cgroup/memory/unimportant/`
+* `/sys/fs/cgroup/cpuacct/unimportant/`
+
+In addition, controllers for one resource group could depend on another in ways that were not
+always obvious.
+
+For example, the `cpu` controller depends on the `cpuacct` controller yet they were treated
+separately.
+
+Many found it confusing for a single process to belong to different nodes in the `cgroup` hierarchy
+across controllers.
+
+The Kernel direction for `cgroup` support is to move toward a unified `cgroup` hierarchy, where the
+per-controller hierarchies are eliminated in favor of hierarchies like the following:
+
+* `/sys/fs/cgroup/important/`
+* `/sys/fs/cgroup/unimportant/`
+
+In a unified hierarchy, a process may only belong to a single node in the `cgroup` tree.
+
+The Kernel direction for `cgroup` management is to promote a single-writer model rather than
+allowing multiple processes to independently write to parts of the file-system.
+
+## cgroup management with systemd and container runtimes
+
+A `slice` corresponds to an inner-node in the `cgroup` file-system hierarchy.
+
+For example, the `system.slice` is represented as follows:
+
+`/sys/fs/cgroup/<controller>/system.slice`
+
+A `slice` is nested in the hierarchy by its naming convention.
+
+For example, the `system-foo.slice` is represented as follows:
+
+`/sys/fs/cgroup/<controller>/system.slice/system-foo.slice/`
+
+A `service` or `scope` corresponds to leaf nodes in the `cgroup` file-system hierarchy.
+
+For example, if the `docker.service` is associated with the `system.slice`, it is
+represented as follows:
+
+`/sys/fs/cgroup/<controller>/system.slice/docker.service/`
+
+To demonstrate the use of `scope` units using the `docker` container runtime, if a
+user launches a container via `docker run -m 100M busybox`, a `scope` will be created
+because the process was not launched by `systemd` itself.  The `scope` is parented by
+the `slice` associated with the launching daemon.
+
+For example:
+
+`/sys/fs/cgroup/<controller>/system.slice/docker-<container-id>.scope`
+
+`systemd` defines a set of slices.  By default, service and scope units are placed in
+`system.slice`, virtual machines and containers registered with `systemd-machined` are
+found in `machine.slice`, and user sessions handled by `systemd-logind` in `user.slice`.
+
+## kubelet cgroup driver
+
+The current implementation of the `kubelet` reads and writes to the `cgroup` file-system
+directly during initial bootstrapping in ways that are not appropriate for `systemd`
+systems.
+
+This proposal proposes that the `kubelet` accept a new flag:
+
+* `--cgroup-driver="":` - cgroup driver used by the kubelet.
```

Comment:
@vishh - do you still see two internal implementations of `ContainerManager` or do you see a single implementation that disables the write path when on `systemd`?  I had assumed it was inevitable that we thought `ContainerManager` may have to do write at some point when we do qos tiers.